File: java/src/sdk/dataobjects/paypal/payflow/FraudResponse.java
package paypal.payflow;

import org.apache.xerces.parsers.DOMParser;
import org.w3c.dom.Document;
import org.w3c.dom.Element;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;
import org.xml.sax.InputSource;

import java.io.ByteArrayInputStream;
import java.util.ArrayList;
import java.util.Hashtable;

/**
 * Container class for response messages specific Fraud Protections Services.
 * <p/>
 * This class contains the fraud protection
 * services related response messages and data objects parsed
 * from the xml data in the fraud response.
 * </p>
 * {@link FpsXmlData}
 */
public final class FraudResponse extends BaseResponseDataObject {

    private Context context;
    private String preFpsMsg;
    private String postFpsMsg;
    private FpsXmlData fpsPreXmlData;
    private FpsXmlData fpsPostXmlData;

    /**
     * Gets the PreFpsMsg parameter.
     *
     * @return - String
     *  Maps to Payflow Parameter: PREFPSMSG
     */
    public String getPreFpsMsg() {
        return preFpsMsg;
    }

    /**
     * Gets the PostFpsMsg parameter.
     *
     * @return - String
     *  Maps to Payflow Parameter: POSTFPSMSG
     */
    public String getPostFpsMsg() {
        return postFpsMsg;
    }

    /**
     * Gets the Fps_PreXmlData parameter.
     *
     * @return - String
     *         <p>Its an itemized list of responses for triggered filters</p>
     *         {@link FpsXmlData}
     */
    public FpsXmlData getFpsPreXmlData() {
        return fpsPreXmlData;
    }

    /**
     * Gets the FpsPostXmlData parameter.
     *
     * @return - String
     *         Gets the FPS Post Xml data message populated in FpsXmlData object.
     *         {@link FpsXmlData}
     *  <p>Maps to Payflow Parameter: FPS_POSTXMLDATA
     */
    public FpsXmlData getFpsPostXmlData() {
        return fpsPostXmlData;
    }


    protected void setContext(Context context) {
        this.context = context;
    }

    protected FraudResponse() {
    }

    private ArrayList parseXmlData(String XmlData) throws Exception {

        DOMParser parser = new DOMParser();
        ByteArrayInputStream byteStream = new ByteArrayInputStream(XmlData.getBytes());
        InputSource source = new InputSource(byteStream);
        ArrayList fraudRuleList = new ArrayList();
        //try
        //{
        parser.parse(source);

        Document xmlDocument = parser.getDocument();

        NodeList ruleList = xmlDocument.getElementsByTagName(PayflowConstants.XML_PARAM_RULE);
        if (ruleList != null) {
            int length = ruleList.getLength();
            if (length > 0) {
                for (int index = 0; index < length; index++) {
                    Node ruleNode = ruleList.item(index);
                    Rule currRule = new Rule();
                    currRule.setNum(Integer.parseInt(ruleNode.getAttributes().getNamedItem(PayflowConstants.XML_PARAM_NUM).getNodeValue()));
                    NodeList childNodes = ruleNode.getChildNodes();
                    int childLength = childNodes.getLength();
                    for (int cindex = 0; cindex < childLength; cindex++) {
                        Element currChild = (Element) childNodes.item(cindex);
                        String Name = currChild.getNodeName();
                        Node child = currChild.getFirstChild();
                        if (PayflowConstants.XML_PARAM_RULEID.equals(Name)) {
                            currRule.setRuleId(child.getNodeValue());
                        } else if (PayflowConstants.XML_PARAM_RULEALIAS.equals(Name)) {
                            currRule.setRuleAlias(child.getNodeValue());
                        } else if (PayflowConstants.XML_PARAM_RULEDESCRIPTION.equals(Name)) {
                            currRule.setRuleDescription(child.getNodeValue());
                        } else if (PayflowConstants.XML_PARAM_ACTION.equals(Name)) {
                            currRule.setAction(child.getNodeValue());
                        } else if (PayflowConstants.XML_PARAM_TRIGGEREDMESSAGE.equals(Name)) {
                            currRule.setTriggeredMessage(child.getNodeValue());
                        } else if (PayflowConstants.XML_PARAM_RULEVENDORPARMS.equals(Name)) {
                            Element ruleVendorParams;
                            ruleVendorParams = currChild;
                            NodeList ruleParamList = ruleVendorParams.getElementsByTagName(PayflowConstants.XML_PARAM_RULEPARAMETER);
                            int ruleParamLength = ruleParamList.getLength();
                            for (int rindex = 0; rindex < ruleParamLength; rindex++) {
                                Node ruleParamNode = ruleParamList.item(rindex);
                                RuleParameter currRuleParam = new RuleParameter();
                                currRuleParam.setNum(Integer.parseInt(ruleParamNode.getAttributes().getNamedItem(PayflowConstants.XML_PARAM_NUM).getNodeValue()));
                                NodeList ruleParamChildNodes = ruleParamNode.getChildNodes();
                                int ruleParamChildLength = ruleParamChildNodes.getLength();
                                for (int rpindex = 0; rpindex < ruleParamChildLength; rpindex++) {
                                    Element currRuleParamChild = (Element) ruleParamChildNodes.item(rpindex);
                                    String ruleParamName = currRuleParamChild.getNodeName();
                                    Node ruleParamChild = currRuleParamChild.getFirstChild();
                                    if (PayflowConstants.XML_PARAM_NAME.equals(ruleParamName)) {
                                        currRuleParam.setName(ruleParamChild.getNodeValue());
                                    } else if (PayflowConstants.XML_PARAM_VALUE.equals(ruleParamName)) {
                                        currRuleParam.setType(currRuleParamChild.getAttribute(PayflowConstants.XML_PARAM_TYPE));
                                        currRuleParam.setValue(ruleParamChild.getNodeValue());
                                    }
                                }
                                currRule.getRuleVendorParms().add(currRuleParam);
                            }
                        }

                    }
                    fraudRuleList.add(currRule);
                }
            }
        }

        //}
        //catch(Exception ex)
        //{
        //	throw ex;
        //}
        return fraudRuleList;
    }


    protected void setParams(Hashtable ResponseHashTable) {
        preFpsMsg = (String) ResponseHashTable.get(PayflowConstants.PARAM_PREFPSMSG);
        postFpsMsg = (String) ResponseHashTable.get(PayflowConstants.PARAM_POSTFPSMSG);
        ResponseHashTable.remove(PayflowConstants.PARAM_PREFPSMSG);
        ResponseHashTable.remove(PayflowConstants.PARAM_POSTFPSMSG);
        SetFpsXmlData(ResponseHashTable);
    }

    private void SetFpsXmlData(Hashtable ResponseHashTable) {
        String XmlData;
        XmlData = (String) ResponseHashTable.get(PayflowConstants.PARAM_FPS_PREXMLDATA);
        fpsPreXmlData = SetRules(XmlData);
        XmlData = (String) ResponseHashTable.get(PayflowConstants.PARAM_FPS_POSTXMLDATA);
        fpsPostXmlData = SetRules(XmlData);
        ResponseHashTable.remove(PayflowConstants.PARAM_FPS_PREXMLDATA);
        ResponseHashTable.remove(PayflowConstants.PARAM_FPS_POSTXMLDATA);
    }

    private FpsXmlData SetRules(String XmlData) {
        FpsXmlData FpsData = new FpsXmlData();
        try {
            if (XmlData != null && XmlData.length() > 0) {
                ArrayList ruleList;

                ruleList = parseXmlData(XmlData);
                if (ruleList != null && ruleList.size() > 0) {
                    FpsData.SetRuleList(ruleList);
                }
            }
        }
        catch (Exception ex) {
            ErrorObject err = new ErrorObject(PayflowConstants.SEVERITY_FATAL, "", "Error occured while parsing: " + ex.toString());
            if (context != null) {
                context.addError(err);
            }
        }
        return FpsData;
    }

}

File: dotNET/PayflowProSDK/PayPal/Payments/DataObjects/FraudResponse.cs
#region "Imports"

using System;
using PayPal.Payments.Common.Utility;
using PayPal.Payments.Common.Exceptions;
using PayPal.Payments.Common;
using System.Collections;
using System.Xml;

#endregion

namespace PayPal.Payments.DataObjects
{
	/// <summary>
	/// Container class for response messages 
	/// specific Fraud Protections Services
	/// </summary>
	/// <remarks>This class contains the fraud protection 
	/// services related response messages and data objects parsed 
	/// from the xml data in the fraud response.
	/// <seealso cref="FpsXmlData"/>
	/// </remarks>
	public sealed class FraudResponse : BaseResponseDataObject
	{
		/// <summary>
		/// Holds PreFpsMsg
		/// </summary>
		private String mPreFpsMsg;

		/// <summary>
		/// Holds PostFpsMsg
		/// </summary>
		private String mPostFpsMsg;

		/// <summary>
		/// Holds Fps Pre Xml Data
		/// </summary>
		private FpsXmlData mFpsPreXmlData;

		/// <summary>
		/// Holds Fps Post Xml Data
		/// </summary>
		private FpsXmlData mFpsPostXmlData;

		#region "Constructors"

		/// <summary>
		/// Constructor for Fraud response.
		/// </summary>
		internal FraudResponse ()
		{

		}


		#endregion

		#region "Properties"

		
		/// <summary>
		/// Gets, Sets  PreFpsMsg
		/// </summary>
		/// <remarks>
		/// Gets the FPS Pre FPS message.
		/// <para>Maps to Payflow Parameter:</para>
		/// <code>PREFPSMSG</code>
		/// </remarks>
		public String PreFpsMsg
		{
			get { return mPreFpsMsg; }
		}

		/// <summary>
		/// Gets, Sets  PostFpsMsg
		/// </summary>
		/// <remarks>
		/// Gets the FPS Post FPS message.
		/// <para>Maps to Payflow Parameter:</para>
		/// <code>POSTFPSMSG</code>
		/// </remarks>
		public String PostFpsMsg
		{
			get { return mPostFpsMsg; }
		}		

		/// <summary>
		/// Gets, Sets  Fps_PreXmlData
		/// </summary>
		/// <remarks>
		/// Gets the FPS Pre Xml data message populated in
		/// FpsXmlData object.
		/// Its an itemized list of responses for trigerred filters
		/// <para>Maps to Payflow Parameter:</para>
		/// <code>FPS_PREXMLDATA</code>
		/// <seealso cref="FpsXmlData"/>
		/// </remarks>
		public FpsXmlData Fps_PreXmlData
		{
			get { return mFpsPreXmlData; }
		}

		/// <summary>
		/// Gets, Sets  Fps_PostXmlData
		/// </summary>
		/// <remarks>
		/// Gets the FPS Post Xml data message populated in
		/// FpsXmlData object.
		/// <para>Maps to Payflow Parameter:</para>
		/// <code>FPS_POSTXMLDATA</code>
		/// <seealso cref="FpsXmlData"/>
		/// </remarks>
		public FpsXmlData Fps_PostXmlData
		{
			get { return mFpsPostXmlData; }
		}

		#endregion

		#region "Functions"

		/// <summary>
		/// Sets the Response params in
		/// response data objects.
		/// </summary>
		///<param name="ResponseHashTable">Response Hash table by ref</param>
		internal void SetParams(ref Hashtable ResponseHashTable)
		{
			try
			{
				
				mPreFpsMsg = (String) ResponseHashTable[PayflowConstants.PARAM_PREFPSMSG];
				mPostFpsMsg = (String) ResponseHashTable[PayflowConstants.PARAM_POSTFPSMSG];
				
				
				ResponseHashTable.Remove(PayflowConstants.PARAM_PREFPSMSG);
				ResponseHashTable.Remove(PayflowConstants.PARAM_POSTFPSMSG);
				
				SetFpsXmlData(ref ResponseHashTable);
			}
			catch (BaseException)
			{
				throw;
			}
			catch (Exception Ex)
			{
				DataObjectException DEx = new DataObjectException(Ex);
				throw DEx;
			}
			//catch
			//{
			//    throw new Exception();
			//}
		}

		/// <summary>
		/// Sets the Fps Xml data
		/// </summary>
		///<param name="ResponseHashTable">Response Hash table by ref</param>
		private void SetFpsXmlData(ref Hashtable ResponseHashTable)
		{
			String XmlData = null;
			try
			{
				XmlData = (String) ResponseHashTable[PayflowConstants.PARAM_FPS_PREXMLDATA];
				mFpsPreXmlData = SetRules(XmlData);
				XmlData = (String) ResponseHashTable[PayflowConstants.PARAM_FPS_POSTXMLDATA];
				mFpsPostXmlData = SetRules(XmlData);
				ResponseHashTable.Remove(PayflowConstants.PARAM_FPS_PREXMLDATA);
				ResponseHashTable.Remove(PayflowConstants.PARAM_FPS_POSTXMLDATA);
			}
			catch (BaseException)
			{
				throw;
			}
			catch (Exception Ex)
			{
				DataObjectException DEx = new DataObjectException(Ex);
				throw DEx;
			}
			//catch
			//{
			//    throw new Exception();
			//}

		}

		/// <summary>
		/// Sets the FPS rules applied.
		/// </summary>
		/// <param name="XmlData">Xml String</param>
		/// <returns>FPS Xml Data object</returns>
		private FpsXmlData SetRules(String XmlData)
		{
			try
			{
				FpsXmlData FpsData = new FpsXmlData();
				if (XmlData != null && XmlData.Length > 0)
				{
					ArrayList RuleList = null;

					RuleList = ParseXmlData(XmlData);
					if (RuleList != null && RuleList.Count > 0)
					{
						FpsData.SetRuleList(RuleList);
					}

				}
				return FpsData;
			}
			catch (BaseException)
			{
				throw;
			}
			catch (Exception Ex)
			{
				DataObjectException DEx = new DataObjectException(Ex);
				throw DEx;
			}
			//catch 
			//{
			//    throw new Exception();
			//}
		}

		/// <summary>
		/// Parses FPS Xml String Data
		/// </summary>
		/// <param name="XmlData">Xml String Data</param>
		/// <returns>Rulelist</returns>
		private ArrayList ParseXmlData(String XmlData)
		{
			try
			{
				ArrayList RuleList = new ArrayList();
				XmlDocument FpsXmlRules = new XmlDocument();
				FpsXmlRules.LoadXml(XmlData);
				XmlNodeList RuleNodeList = FpsXmlRules.GetElementsByTagName(PayflowConstants.XML_PARAM_RULE);
				if (RuleNodeList != null && RuleNodeList.Count > 0)
				{
					foreach (XmlNode RuleNode in RuleNodeList)
					{
						String TempValue;
						Rule FraudRule = new Rule();
						TempValue = RuleNode.Attributes.GetNamedItem(PayflowConstants.XML_PARAM_NUM).Value;
						if (TempValue != null)
						{
							FraudRule.Num = Int32.Parse(TempValue);
						}
						FraudRule.RuleId = RuleNode.SelectSingleNode(PayflowConstants.XML_PARAM_RULEID).InnerText;
						FraudRule.RuleAlias = RuleNode.SelectSingleNode(PayflowConstants.XML_PARAM_RULEALIAS).InnerText;
						FraudRule.RuleDescription = RuleNode.SelectSingleNode(PayflowConstants.XML_PARAM_RULEDESCRIPTION).InnerText;
						FraudRule.Action = RuleNode.SelectSingleNode(PayflowConstants.XML_PARAM_ACTION).InnerText;
						FraudRule.TriggeredMessage = RuleNode.SelectSingleNode(PayflowConstants.XML_PARAM_TRIGGEREDMESSAGE).InnerText;
						XmlNode RuleVendorParamNode = RuleNode.SelectSingleNode(PayflowConstants.XML_PARAM_RULEVENDORPARMS);
						if (RuleVendorParamNode != null)
						{
							XmlNodeList RuleParamList = RuleVendorParamNode.SelectNodes(PayflowConstants.XML_PARAM_RULEPARAMETER);
							if (RuleParamList != null)
							{
								foreach (XmlNode RuleParamNode in RuleParamList)
								{
									RuleParameter RuleParam = new RuleParameter();
									String TempValue1;
									TempValue1 = RuleParamNode.Attributes.GetNamedItem(PayflowConstants.XML_PARAM_NUM).Value;
									if (TempValue1 != null)
										RuleParam.Num = Int32.Parse(TempValue);
									RuleParam.Name = RuleParamNode.SelectSingleNode(PayflowConstants.XML_PARAM_NAME).InnerText;
									RuleParam.Type = RuleParamNode.SelectSingleNode(PayflowConstants.XML_PARAM_VALUE).Attributes.GetNamedItem(PayflowConstants.XML_PARAM_TYPE).Value;
									RuleParam.Value = RuleParamNode.SelectSingleNode(PayflowConstants.XML_PARAM_VALUE).InnerText;
									FraudRule.RuleVendorParms.Add(RuleParam);
								}
							}
						}
						RuleList.Add(FraudRule);
					}
				}

				return RuleList;
			}
			catch (BaseException)
			{
				throw;
			}
			catch (Exception Ex)
			{
				ErrorObject Error = new PayPal.Payments.Common.ErrorObject("Error While parsing XmlData", Ex.Message);
				DataObjectException DEx = new DataObjectException(Error);
				throw DEx;
			}
			//catch
			//{
			//    ErrorObject Error = new PayPal.Payments.Common.ErrorObject("Error While parsing XmlData", "");
			//    DataObjectException DEx = new DataObjectException(Error);
			//    throw DEx;
			//}
		}

		#endregion
	}

}